### PR TITLE
Restructure Codebase

### LINKED ISSUE #147: Useless `Partial` in type definition
I don't know why `Partial<>` is used for the types of params.
```typescript
declare function deepmerge<T1, T2>(x: Partial<T1>, y: Partial<T2>, options?: deepmerge.Options): T1 & T2;
```

The `Partial<>` causes a difference between static typing and actual values in runtime.
```typescript
interface T {
  a: string;
  b?: string;
}
const t1: T = { a: "foo" };
const t2: T = { a: "bar" };

// The inferred type of `result` is `{a: string, b: string}`,
// but the actual value is `{a: "bar"}` without prop `b`.
const result = merge(t1, t2);

// tsc doesn't throw any compile errors, but a runtime error is thrown.
// TypeError: Cannot read property 'toLowerCase' of undefined
console.log(result.b.toLowerCase());
```

I think the `Partial<>` should be just removed.
```typescript
declare function deepmerge<T1, T2>(x:T1, y: T2, options?: deepmerge.Options): T1 & T2;
```

### LINKED ISSUE #161: Drop ES5
I'm ready to convert to modern JS.

A note should probably be added to the readme.

### LINKED ISSUE #179: d.ts type definition error
image

⬆️ the type of `t4.a` should be a `number` type, not `never`.

### LINKED ISSUE #180: Test the type definition with tsd
tsd: https://github.com/SamVerschueren/tsd#readme

It would be great to be able to make more explicit assertions about the type definition.

### PR DESCRIPTION
**Pull request by RebeccaStevens**
- Refactor code into multiple files
- Switch to using TypeScript
- Update dev dependencies

fixes #180
fixes #179
fixes #161
fixes #147

closes #213
closes #211
closes #200
closes #198
closes #188 
closes #181
closes #173 
closes #148

---
This change is [Reviewable](https://reviewable.io/reviews/tehshrike/deepmerge/215)

### CODE REVIEW

**Comment by RebeccaStevens** [comment]
@TehShrike Could you provide me with the eslint config you are using locally for this project?

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+[*]
+indent_style = tab
+end_of_line = lf
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = true
+
+[*.md]
+indent_style = space
```
Comment (TehShrike) [inline .editorconfig:11]: I do tabs/trim-trailing-whitespace in Markdown too

File: .editorconfig [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+indent_size = 2
+trim_trailing_whitespace = false
+
+[{package.json,package-lock.json}]
+indent_style = space
```
Comment (TehShrike) [inline .editorconfig:16]: cursed npm-enforced defaults

File: .npmignore [hunk 3]
```diff
@@ -1,3 +0,0 @@
-test/
```
Comment (TehShrike) [inline .npmignore:1]: I don't know if there's any value in pushing the extra bytes of test files to everyone via npm

**Review by TehShrike — COMMENTED** [review]
Some quick ancillary notes, I'll circle back and read the actual implementation more carefully later.

One other picky note: I really dislike javadoc/tsdoc/jsdoc function comments, if any of the words in them actually add something, I would rather just move those words into the function names.  e.g. if "A function that determins if a type is mergable" is more meaningful, then we should rename `IsMergeable` to `IsTypeMergeable` or something.

File: .editorconfig (hunk 1 above)
Comment (RebeccaStevens) [reply .editorconfig:11]: My reason for switching to spaces in markdown files was just to ensure they render the same regardless of what render is being used - GitHub/npm/other.
But I don't have a strong opinion on this either way. I put this change its own commit so it can easily be dropped.

The reason for not trimming trailing whitespace at the end of markdown files is because two spaces at the end of a line is part of the markdown spec (meaning add a line break here). Official markdown doesn't treat a single line break any different to a single space when it comes to rendering. However, GitHub-flavored markdown, does treat single line break as an actual line break.
If you want to continue trimming trailing whitespace I'm ok with that but I'd recommend not doing so.

File: .npmignore (hunk 3 above)
Comment (RebeccaStevens) [reply .npmignore:1]: I removed the `.npmignore` file in favor of using the `"files"` field in the `package.json` file.

The `"files"` field works as a whitelist as apposed to a blacklist like the `.npmignore` file.

I've got the files set to the following (2 folders and 3 files (plus the package.json file itself)):

```json
"files": [
  "dist",
  "types",
  "changelog.md",
  "license.txt",
  "readme.md"
],
```

You can run `npm pack` to create the package locally so you can see what's included in it.

File: .editorconfig (hunk 1 above)
Comment (TehShrike) [reply .editorconfig:11]: Whitespace at the end of the line can be meaningful in Markdown, but I never use that line-break functionality in documentation

File: .npmignore (hunk 3 above)
Comment (TehShrike) [reply .npmignore:1]: Ah cool, yeah, I like `files` for that

File: .eslintrc.json [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+				],
+				"@typescript-eslint/no-use-before-define": [
+					"error",
+					{
+						"functions": false,
+						"variables": true
+					}
+				],
+
+				"@typescript-eslint/prefer-for-of": "warn",
```
Comment (TehShrike) [inline .eslintrc.json:177]: I detest for-of loop, personally.

I don't want to get bogged down in a full audit of this eslintrc, maybe it could be pulled out to its own PR?

**Review by TehShrike — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,49 @@
+name: CI
```
Comment (TehShrike) [inline .github/workflows/ci.yml:1]: why move away from Travis in the same PR as all these other changes?

File: .editorconfig (hunk 1 above)
Comment (TehShrike) [reply .editorconfig:11]: But the indent_style should definitely be tab 😅

File: changelog.md [hunk 6]
```diff
@@ -1,5 +1,9 @@
 # [5.0.0](https://github.com/TehShrike/deepmerge/releases/tag/v5.0.0)
 
+- Breaking: Endpoint are now exported in esm style [#215](https://github.com/TehShrike/deepmerge/pull/215)
```
Comment (TehShrike) [inline changelog.md:3]: It's a bit tough for folks to go check out the PR to see exactly what the change was and run into
image

File: readme.md [hunk 7]
```diff
@@ -2,56 +2,56 @@
...
-> UMD bundle is 723B minified+gzipped
+> UMD bundle is 792B minified+gzipped
 
 ## Getting Started
 
 ### Example Usage
+
 <!--js
-const merge = require('./')
+const { default: deepmerge, deepmergeAll } = require('.')
```
Comment (TehShrike) [inline readme.md:12]: node has always been a first-class consumer as far as deepmerge is concerned – I'm not okay with moving away from

```js
const deepmerge = require('deepmerge')
```

for node.js/CommonJS
